Pick up item in hand and shooty-tooty

Alexamakans opened the pull request:
If you're looking at an object with the Ammunition tag and right-click (Fire2) you pick it up.
You can then use left-click (Fire1) to shoot.

Object is scaled down when held in hand, then the scale is restored.
Object also ignores collision with the player for one second after shooting.
Object is destroyed 5 seconds after shooting, or on impact, whichever happens first.

Alexamakans committed:
Added Ammunition tag and Ignore Player layer

Alexamakans committed:
Added Ammunition component

Alexamakans committed:
Added re-enable collisions and revert to original scale in OnShoot

Alexamakans committed:
Added PickupAmmunition component

Alexamakans committed:
Changed PlayerShooting component to use the new PickupAmmunition component

Alexamakans committed:
Updated Player prefab to use the new PickupAmmunition component

Alexamakans committed:
Updated scene

applejag commented on Assets/Scripts/Ammunition.cs:
```diff
@@ -0,0 +1,57 @@
+    public Transform hand;
+    public Vector3 inHandSize = new Vector3(0.2f, 0.2f, 0.2f);
+
+    public Vector3 originalScale;
```
Does this need to be public? Especially since `_originalLayer` is made private.

applejag commented on Assets/Scripts/Ammunition.cs:
```diff
@@ -0,0 +1,57 @@
+    public Vector3 originalScale;
+
+    private int _originalLayer;
+    private Rigidbody _body;
```
Suggest making this a public field and setting it in `void Reset()`.

applejag commented on Assets/Scripts/PickupAmmunition.cs:
```diff
@@ -0,0 +1,34 @@
+                && !collider.GetComponent<Ammunition>())
+            {
+                var go = collider.gameObject;
+                ammunition = go.AddComponent<Ammunition>();
```
Adding and removing components through code has led me personally through rabbitholes of debugging.
Suggest here to instead always have the Ammunition component on there, and call something like a "Pickup" method that does what the Start method does now.

Perhaps something like:

```cs
var ammunition = collider.gameObject.GetComponent<Ammunition>();
if (ammunition) {
    ammunition.Pickup(inHandSize);
}
```

Could probably discard the ammunition tag then as well

applejag commented on Assets/Scripts/Ammunition.cs:
```diff
@@ -0,0 +1,57 @@
+
+public class Ammunition : MonoBehaviour
+{
+    public Transform hand;
```
This `hand` reference seems to only be used as a boolean. Suggest replacing it with `bool isPickedUp` instead.

applejag commented on Assets/Scripts/PickupAmmunition.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        if (!ammunition
+            && Input.GetButtonDown("Fire2")
+            && Physics.Raycast(Camera.main.ScreenPointToRay(Input.mousePosition), out var hitInfo, 2f, LayerMask.NameToLayer("Player")))
```
Suggest extracting some of these values as public fields. Namely the distance and layer mask.

applejag commented on Assets/Scripts/PickupAmmunition.cs:
```diff
@@ -0,0 +1,34 @@
+using UnityEngine;
+
+public class PickupAmmunition : MonoBehaviour
```
nit: I find it difficult to see the distinction in the names `Ammunition` and `PickupAmmunition`. Suggest something like this (especially trying to keep the `Player-` prefix on player components):

- `Ammunition` &rarr; `Pickupable` (or keep `Ammunition`, it's quite good)
- `PickupAmmunition` &rarr; `PlayerInventory` or `PlayerWeapon`

applejag commented on Assets/Scripts/PlayerShooting.cs:
```diff
@@ -16,28 +22,26 @@ void Update()
-        else
+
+        var ammoBody = pickupAmmunition.ammunition.GetComponent<Rigidbody>();
+        if (!ammoBody)
         {
-            body.AddForce(fireFrom.forward * fireForce);
+            Debug.LogWarning("Projectile does not have a Rigidbody.", this);
+            return;
         }
```
`Ammunition` component already has a reference to its Rigidbody. Think you can take use of that here.

applejag commented on ProjectSettings/TagManager.asset:
```diff
@@ -5,15 +5,16 @@ TagManager:
   - TransparentFX
   - Ignore Raycast
-  - 
+  - Player
```
Do we want to merge the `Player` and `Hands` layers?

Alexamakans committed:
Made originalScale private

Alexamakans committed:
Moved Start code to new method PickUp and made _body a public field

Alexamakans committed:
Removed Ammunition tag

Alexamakans committed:
Extracted distance and layer mask as public fields

Alexamakans committed:
Removed hand reference from Ammunition and moved Ammunition.Update code to PickupAnimation

Alexamakans committed:
Renamed PickupAmmunition -> PlayerInventory

Alexamakans committed:
Changed to use public field Ammunition.body instead of GetComponent call

Alexamakans committed:
Combined Hands and Player layers

Alexamakans committed:
Added Ammunition.canBePickedUp boolean

applejag commented on Assets/Scripts/Ammunition.cs:
```diff
@@ -2,43 +2,35 @@
 
-        _body = GetComponent<Rigidbody>();
-        _body.detectCollisions = false;
+        body.detectCollisions = false;
```
Might want to make it kinematic as well to disable stuff like gravity, or just manually disabling gravity as well. Or is gravity handled elsewhere?

applejag commented on Assets/Scripts/PlayerInventory.cs:
```diff
@@ -0,0 +1,40 @@
+    public float pickUpRange;
+    public Transform hand;
+    public Vector3 inHandSize = new Vector3(0.2f, 0.2f, 0.2f);
+    public GameObject projectileHitParticles;
```
I see that OnShoot accepted a game object before as well, but why do we really need to pass this reference along down the line?

If it's a prefab, it should be enough to just assign it in the Ammunition component directly.

Alexamakans replied:
Leftover from when the Ammunition component was added to the object in the PlayerInventory (prev. PickupAmmunition) component.

Will change.

applejag commented on Assets/Scripts/PlayerShooting.cs:
```diff
@@ -33,13 +33,7 @@ void FireProjectile()
-        {
-            Debug.LogWarning("Projectile does not have a Rigidbody.", this);
-            return;
-        }
```
I think it's wise to keep warnings like these, or if it's more crucial then promoting it to a LogError instead

Alexamakans committed:
Changed Ammunition.body to be kinematic when picked up

This change removed the need to constantly set the velocity to zero in PlayerInventory.Update

Alexamakans committed:
Readded Debug.LogWarning call to PlayerShooting if no Rigidbody on pickupAmmunition.ammunition

Alexamakans committed:
Renamed PlayerShooting.pickupAmmunition -> playerInventory to match component name

Alexamakans committed:
Moved impactParticles field from PlayerInventory to Ammunition

Alexamakans committed:
Changed to not Destroy(this) in Ammunition.UnignorePlayer

Alexamakans committed:
Merged main into ammunition
